File: delivery-multiagente/backend/src/routes/driver.ts
import { Router } from 'express';
import { authMiddleware, requireRole } from '../middleware/auth';
import { DeliveryAgent } from '../agents/DeliveryAgent';
import { prisma } from '../lib/prisma';

const router = Router();
const deliveryAgent = new DeliveryAgent();

router.use(authMiddleware, requireRole('ADMIN', 'REPARTIDOR'));

// Pedidos en reparto
router.get('/active', async (_req, res, next) => {
  try { res.json(await deliveryAgent.getActiveDeliveries()); } catch (err) { next(err); }
});

// Pedidos listos para asignar
router.get('/ready-orders', async (_req, res, next) => {
  try {
    const orders = await prisma.order.findMany({
      where: { status: 'LISTO_PARA_REPARTO' },
      include: { items: { include: { product: true } } },
    });
    res.json(orders);
  } catch (err) { next(err); }
});

// Confirmar entrega
router.post('/orders/:id/confirm-delivery', async (req, res, next) => {
  try {
    const result = await deliveryAgent.confirmDelivery(req.params.id);
    res.json(result);
  } catch (err) { next(err); }
});

// Pedidos completados (historial del día)
router.get('/completed', async (_req, res, next) => {
  try {
    const orders = await prisma.order.findMany({
      where: { status: { in: ['ENTREGADO', 'CERRADO'] } },
      include: { items: { include: { product: true } }, driver: true },
      orderBy: { updated_at: 'desc' },
      take: 30,
    });
    res.json(orders);
  } catch (err) { next(err); }
});

// Repartidores disponibles
router.get('/available-drivers', async (_req, res, next) => {
  try { res.json(await deliveryAgent.getAvailableDrivers()); } catch (err) { next(err); }
});

export default router;

File: delivery-langgraph/el_trujillano/api/routers/driver.py
"""Router de repartidor (roles ADMIN, REPARTIDOR). Determinista."""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import joinedload, selectinload

from ... import order_flow
from ...db.database import get_session
from ...db.models import Driver, Order
from ...estados import CERRADO, EN_REPARTO, ENTREGADO, LISTO_PARA_REPARTO
from ..security import require_role
from ..serializers import driver_to_dict, order_to_dict

router = APIRouter(dependencies=[Depends(require_role("ADMIN", "REPARTIDOR"))])

_OPTS = (selectinload(Order.items), joinedload(Order.driver))


@router.get("/active")
def active_deliveries():
    with get_session() as session:
        orders = (
            session.query(Order).options(*_OPTS).filter(Order.estado == EN_REPARTO).all()
        )
        return [order_to_dict(o) for o in orders]


@router.get("/ready-orders")
def ready_orders():
    with get_session() as session:
        orders = (
            session.query(Order).options(*_OPTS).filter(Order.estado == LISTO_PARA_REPARTO).all()
        )
        return [order_to_dict(o) for o in orders]


@router.get("/completed")
def completed():
    with get_session() as session:
        orders = (
            session.query(Order)
            .options(*_OPTS)
            .filter(Order.estado.in_([ENTREGADO, CERRADO]))
            .order_by(Order.updated_at.desc())
            .limit(30)
            .all()
        )
        return [order_to_dict(o) for o in orders]


@router.post("/orders/{order_id}/confirm-delivery")
def confirm_delivery(order_id: int):
    """Repartidor confirma entrega: EN_REPARTO -> ENTREGADO -> CERRADO (crea encuesta)."""
    try:
        order_flow.marcar_entregado(order_id)
        return order_flow.cerrar(order_id)
    except Exception as e:
        raise HTTPException(status_code=409, detail=str(e))


@router.get("/available-drivers")
def available_drivers():
    with get_session() as session:
        drivers = session.query(Driver).filter(Driver.disponible.is_(True)).all()
        return [driver_to_dict(d) for d in drivers]
